source/extensions/path/pattern_template_lib/pattern_template.cc

```diff
@@ -0,0 +1,171 @@
+namespace PatternTemplate {
+
+using PatternTemplateInternal::ParsedUrlPattern;
+
+#ifndef SWIG
+// Silence warnings about missing initializers for members of LazyRE2.
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+#endif
+
+inline re2::StringPiece toStringPiece(absl::string_view text) { return {text.data(), text.size()}; }
+
+absl::StatusOr<std::string> convertURLPatternSyntaxToRegex(absl::string_view url_pattern) {
+
+  absl::StatusOr<ParsedUrlPattern> status =
+      PatternTemplateInternal::parseURLPatternSyntax(url_pattern);
+  if (!status.ok()) {
+    return status.status();
+  }
+  struct ParsedUrlPattern pattern = *std::move(status);
+  return PatternTemplateInternal::toRegexPattern(pattern);
+}
+
+absl::StatusOr<std::vector<RewritePatternSegment>>
+parseRewritePatternHelper(absl::string_view pattern) {
+  std::vector<RewritePatternSegment> result;
+
+  // Don't allow contiguous '/' patterns.
+  static const LazyRE2 invalid_regex = {"^.*//.*$"};
+  if (RE2::FullMatch(toStringPiece(pattern), *invalid_regex)) {
+    return absl::InvalidArgumentError("Invalid rewrite literal pattern");
+  }
+
+  // The pattern should start with a '/' and thus the first segment should
+  // always be a literal.
+  if (pattern.empty() || pattern[0] != '/') {
+    return absl::InvalidArgumentError("Invalid rewrite variable placement");
+  }
+  while (!pattern.empty()) {
+    std::vector<absl::string_view> segments1 = absl::StrSplit(pattern, absl::MaxSplits('{', 1));
+    if (!segments1[0].empty()) {
+      if (!PatternTemplateInternal::isValidRewriteLiteral(segments1[0])) {
+        return absl::InvalidArgumentError("Invalid rewrite literal pattern");
+      }
+      result.emplace_back(segments1[0], RewriteStringKind::KLiteral);
+    }
+
+    if (segments1.size() < 2) {
+      // No more variable replacement, done.
+      break;
+    }
+
+    std::vector<absl::string_view> segments2 =
+        absl::StrSplit(segments1[1], absl::MaxSplits('}', 1));
+    if (segments2.size() < 2) {
+      return absl::InvalidArgumentError("Unmatched variable bracket");
+    }
+    pattern = segments2[1];
+
+    if (!PatternTemplateInternal::isValidIndent(segments2[0])) {
+      return absl::InvalidArgumentError("Invalid variable name");
+    }
+    result.emplace_back(segments2[0], RewriteStringKind::KVariable);
+  }
+  return result;
+}
+
+absl::StatusOr<envoy::extensions::pattern_template::PatternTemplateRewriteSegments>
+parseRewritePattern(absl::string_view pattern, absl::string_view capture_regex) {
+  envoy::extensions::pattern_template::PatternTemplateRewriteSegments parsed_pattern;
+  RE2 regex = RE2(toStringPiece(capture_regex));
+  if (!regex.ok()) {
+    return absl::InternalError(regex.error());
+  }
+
+  absl::StatusOr<std::vector<RewritePatternSegment>> status = parseRewritePatternHelper(pattern);
+  if (!status.ok()) {
+    return status.status();
+  }
+  std::vector<RewritePatternSegment> processed_pattern = *std::move(status);
```

Comment:
Does this actually elide copying the vector out of `status`? I could be wrong, but It seems to me that `std::move(status)` move the vector to a temporary StatusOr and then calling `operator*` on it gets you a `std::vector<RewritePatternSegment>&` which is copied into processed_pattern.

Maybe just make this line `auto& processed_pattern = *status`?

I think there are a number of other places with this `*std::move(x)` pattern that seem suspicious to me.

Reply:
Or, I guess since it's `T&& operator*() &&` this is being moved all the way into the new variable. Still I think the shorter version is easier to read.

Reply:
I think I am going to keep as is. This comes from internal implementation.